### PR TITLE
Add spire version

### PR DESCRIPTION
**Pull request by ed1d1a8d**
Addresses https://github.com/sipb/homeworld/issues/141.

### CODE REVIEW

**Comment by ed1d1a8d** [comment]
The git version is only displayed if homeworld-admin-tools/build-package.sh is called. Calling inner-build does not yield the git version.

File: building/build-debs/homeworld-admin-tools/build-package.sh [hunk 1]
```diff
@@ -1,6 +1,28 @@
 #!/bin/bash
-set -eu
+set -e -u
+rm -f resources/GIT_VERSION
```
Comment (celskeggs) [inline building/build-debs/homeworld-admin-tools/build-package.sh]: There's no need to remove this file explicitly; ` >resources/GIT_VERSION` will overwrite it.

File: building/build-debs/homeworld-admin-tools/build-package.sh [hunk 2]
```diff
@@ -1,6 +1,28 @@
...
+set +e
+git diff-index --quiet HEAD --
+GIT_DIFF_RETVAL=$?
+set -e
+
+GIT_ROOT="$(git rev-parse --show-toplevel)"
+set +e
+u="$(git ls-files --exclude-standard --others "${GIT_ROOT}")" && test -z "$u"
+GIT_LS_RETVAL=$?
+set -e
```
Comment (celskeggs) [inline building/build-debs/homeworld-admin-tools/build-package.sh]: This seems rather complicated. Maybe something from https://unix.stackexchange.com/a/155077 might be easier?

File: building/build-debs/homeworld-admin-tools/debian/rules [hunk 3]
```diff
@@ -17,11 +17,11 @@ include /usr/share/dpkg/default.mk
 #export DEB_LDFLAGS_MAINT_APPEND = -Wl,--as-needed
 
 override_dh_auto_build:
-	VERSION="${DEB_VERSION_UPSTREAM}" ./inner-build.sh
+	DEB_VERSION="${DEB_VERSION_UPSTREAM}" ./inner-build.sh
```
Comment (celskeggs) [inline building/build-debs/homeworld-admin-tools/debian/rules]: Is it necessary to rename this? It breaks consistency with the other debian/rules files.

File: building/build-debs/homeworld-admin-tools/build-package.sh [hunk 4]
```diff
@@ -1,6 +1,28 @@
...
+then
+    echo -n "-dirty" >>resources/GIT_VERSION
+fi
+echo >>resources/GIT_VERSION
+
 source ../common/package-build-helpers.sh
 
 upstream "debian-9.2.0-amd64-mini.iso"
 build
+rm resources/GIT_VERSION
```
Comment (celskeggs) [inline building/build-debs/homeworld-admin-tools/build-package.sh]: Might be worth doing this in a trap handler, so that it happens even if the build fails.

File: building/build-debs/homeworld-admin-tools/src/__main__.py [hunk 5]
```diff
@@ -28,6 +29,7 @@
     "infra": infra.main_command,
     "seq": seq.main_command,
     "deploy": deploy.main_command,
+    "version": command.wrap("display version info", version.display_version)
```
Comment (celskeggs) [inline building/build-debs/homeworld-admin-tools/src/__main__.py]: For the sake of consistency and maintaining abstraction boundaries, it makes more sense to have `version` export a `main_command` attribute that equals `command.wrap("display version info", display_version)` instead of this.

File: building/build-debs/homeworld-admin-tools/src/version.py [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import os
+from resources import get_resource
+
+
+def display_version():
+    deb_version = get_resource("DEB_VERSION").decode("utf-8").rstrip()
+    print("Debian package version:", deb_version)
+
+    git_version = get_resource("GIT_VERSION").decode("utf-8").rstrip()
+    print("Build commit hash:", git_version)
```
Comment (celskeggs) [inline building/build-debs/homeworld-admin-tools/src/version.py]: We generally don't bother with specifying the codec when we call `.decode()`. Do we have a particular new reason to do that?

Also, does this correctly handle the case where we build directly with inner_build.sh and resources/GIT_VERSION is missing?

File: building/build-debs/homeworld-admin-tools/build-package.sh (hunk 1 above)
Comment (ed1d1a8d) [reply building/build-debs/homeworld-admin-tools/build-package.sh]: We either need to remove it or gitignore `resources/GIT_VERSION` since we are checking if the working directory is clean immediately after. Which is preferable?

File: building/build-debs/homeworld-admin-tools/debian/rules (hunk 3 above)
Comment (ed1d1a8d) [reply building/build-debs/homeworld-admin-tools/debian/rules]: Not necessary. It just makes the variable naming in inner-build.sh nicer. Do you want me to change it back?

File: building/build-debs/homeworld-admin-tools/src/__main__.py (hunk 5 above)
Comment (ed1d1a8d) [reply building/build-debs/homeworld-admin-tools/src/__main__.py]: Will change.

File: building/build-debs/homeworld-admin-tools/build-package.sh (hunk 2 above)
Comment (ed1d1a8d) [reply building/build-debs/homeworld-admin-tools/build-package.sh]: WiIl simplify.

File: building/build-debs/homeworld-admin-tools/src/version.py (hunk 6 above)
Comment (ed1d1a8d) [reply building/build-debs/homeworld-admin-tools/src/version.py]: Regarding your first point: I did not realize decode defaulted to utf-8. Will change the code to not specify codec.

Comment (ed1d1a8d) [reply building/build-debs/homeworld-admin-tools/src/version.py]: Regarding your second point: the code crashes when building only with inner-build.sh. What should be the behavior when just building with inner-build? Should it behave the same as outer-build?

**Comment by ed1d1a8d** [comment]
What should be the behavior when just building with inner-build? Should it behave the same as build-package?

File: building/build-debs/homeworld-admin-tools/build-package.sh (hunk 1 above)
Comment (celskeggs) [reply building/build-debs/homeworld-admin-tools/build-package.sh]: gitignoring it seems like the right solution. We wouldn't want to check it in, anyway.

File: building/build-debs/homeworld-admin-tools/debian/rules (hunk 3 above)
Comment (celskeggs) [reply building/build-debs/homeworld-admin-tools/debian/rules]: Yeah -- if you want to submit a separate change to improve the naming, that'd be good, but perhaps not in this PR.

File: building/build-debs/homeworld-admin-tools/src/version.py (hunk 6 above)
Comment (celskeggs) [reply building/build-debs/homeworld-admin-tools/src/version.py]: Well, it should *definitely* not crash! Especially not delayed until runtime like that. A long-distance crash caused by building a different way is an unfortunate bug.

We also do call ./inner-build.sh directly during testing, so we should definitely support this condition. We should probably populate the resource in all cases.

With regards to the debian version, it probably makes more sense to only populate it when using ./build_package.sh, and then say "this was not built as part of a debian package" if ./inner_build.sh was used, either by detecting the lack of the resource or by sticking that string into the file.

It's probably worth, in any case, making sure that there's a clear error message if either of these is missing. It might be clear enough to just say "GIT_VERSION cannot be found" or something, which might already be the error message, but if the error message is more confusing than that, it might be worth adding extra code for a better message.

**Comment by celskeggs** [comment]
Well, not *exactly* the same, but it should definitely succeed and build a working binary. See more specific comments on the review itself.

Once you've addressed the comments, you should assign this PR back to me so that I know I need to take care of it.

File: building/build-debs/homeworld-admin-tools/build-package.sh [hunk 7]
```diff
@@ -15,4 +19,3 @@ source ../common/package-build-helpers.sh
```
Comment (celskeggs) [inline building/build-debs/homeworld-admin-tools/build-package.sh]: These changes should probably be stuck below sourcing of package-build-helpers.sh.

File: building/build-debs/homeworld-admin-tools/build-package.sh [hunk 8]
```diff
@@ -2,23 +2,13 @@
...
-git diff-index --quiet HEAD --
-GIT_DIFF_RETVAL=$?
-set -e
-
-GIT_ROOT="$(git rev-parse --show-toplevel)"
-set +e
-u="$(git ls-files --exclude-standard --others "${GIT_ROOT}")" && test -z "$u"
-GIT_LS_RETVAL=$?
+u="$(git status --porcelain)" && test -z "$u"
+GIT_DIRTY=$?
```
Comment (celskeggs) [inline building/build-debs/homeworld-admin-tools/build-package.sh]: You realize you can just set `u` here and then use `test -z "$u"` down below (or its negation, perhaps) instead of comparing the retval, right?

**Comment by ed1d1a8d** [comment]
I don't think I can edit the assignees list.

File: building/build-debs/homeworld-admin-tools/build-package.sh [hunk 9]
```diff
@@ -1,5 +1,20 @@
...
+trap cleanup EXIT
+cleanup
+
+set +e
+u="$(git status --porcelain)" && test -z "$u"
+GIT_DIRTY=$?
+set -e
+
+git rev-parse HEAD | tr -d '\n' >resources/GIT_VERSION
+[[ $GIT_DIRTY -ne 0 ]] && echo -n "-dirty" >>resources/GIT_VERSION
```
Comment (celskeggs) [inline building/build-debs/homeworld-admin-tools/build-package.sh]: You realize you can just set u above and then use test -z "$u" down here (or its negation, perhaps) instead of comparing the retval, right?

File: building/build-debs/homeworld-admin-tools/build-package.sh (hunk 7 above)
Comment (ed1d1a8d) [reply building/build-debs/homeworld-admin-tools/build-package.sh]: Just wondering, why is this the best practice?

File: building/build-debs/homeworld-admin-tools/build-package.sh (hunk 9 above)
Comment (ed1d1a8d) [reply building/build-debs/homeworld-admin-tools/build-package.sh]: Yes, that sounds like a better idea.

File: building/build-debs/homeworld-admin-tools/build-package.sh (hunk 7 above)
Comment (celskeggs) [reply building/build-debs/homeworld-admin-tools/build-package.sh]: Good question. It's because source is kind of like import, so you do it at the top of the file for the same reason you usually do your imports at the top of a python file: the set of imports informs many things about what the file does, so by putting them at the top, the outline of what the file is doing becomes clearer to the reader.

**Comment by celskeggs** [comment]
You should have access to edit the assignees list now.

**Comment by ed1d1a8d** [comment]
I can confirm that I have access. Thanks.

File: building/build-debs/homeworld-admin-tools/resources/__init__.py [hunk 10]
```diff
@@ -2,9 +2,11 @@
 
 
 def get_resource(name: str) -> bytes:
-    b = pkgutil.get_data(__package__, name)
+    try:
```
Comment (ed1d1a8d) [inline building/build-debs/homeworld-admin-tools/resources/__init__.py]: Catches error when file not found. Per https://docs.python.org/3.5/library/pkgutil.html.
Should this go in a separate commit, since it is not directly related to spire version?

**Comment by ed1d1a8d** [comment]
Works for both build-package and inner-build.

File: building/build-debs/homeworld-admin-tools/inner-build.sh [hunk 11]
```diff
@@ -3,6 +3,21 @@ set -e -u
 
 rm -rf spire spire.zip src/resources src/__pycache__ resources/__pycache__
 cp -RT resources src/resources
+
+if [[ -v VERSION ]]
+then
+    echo "$VERSION" >src/resources/DEB_VERSION
+else
+    echo "Spire was not built as part of a Debian package." >src/resources/DEB_VERSION
```
Comment (celskeggs) [inline building/build-debs/homeworld-admin-tools/inner-build.sh]: Probably worth having a less verbose string would be good?

This is long:

    Debian package version: Spire was not built as part of a Debian package.

Compared to something like this:

    Debian package version: not a debian build

File: building/build-debs/homeworld-admin-tools/resources/__init__.py (hunk 10 above)
Comment (celskeggs) [reply building/build-debs/homeworld-admin-tools/resources/__init__.py]: Because it's a change involved in implementing `spire version`, it makes sense to include it in the same PR. You can choose whether it's cleaner in a separate commit within the PR -- whatever seems right.

File: building/build-debs/homeworld-admin-tools/inner-build.sh [hunk 12]
```diff
@@ -3,6 +3,21 @@ set -e -u
...
+fi
+
+if [[ ! -e src/resources/GIT_VERSION ]]
+then
+    git rev-parse HEAD | tr -d '\n' >src/resources/GIT_VERSION
+    [[ -n "$(git status --porcelain)" ]] && echo -n "-dirty" >>src/resources/GIT_VERSION
+    echo >>src/resources/GIT_VERSION
+fi
```
Comment (celskeggs) [inline building/build-debs/homeworld-admin-tools/inner-build.sh]: I assume that the reason that this code is duplicated is because the git repository can't be accessed from within the build chroot?

In any case, you should make sure that inner_build.sh cleans up GIT_VERSION.

File: building/build-debs/homeworld-admin-tools/resources/__init__.py (hunk 10 above)
Comment (celskeggs) [reply building/build-debs/homeworld-admin-tools/resources/__init__.py]: Because it's a change involved in implementing spire version, it makes sense to include it in the same PR. You can choose whether it's cleaner in a separate commit within the PR -- whatever seems right.

File: building/build-debs/homeworld-admin-tools/inner-build.sh (hunk 11 above)
Comment (ed1d1a8d) [reply building/build-debs/homeworld-admin-tools/inner-build.sh]: Will shorten to what you have there.

File: building/build-debs/homeworld-admin-tools/inner-build.sh (hunk 12 above)
Comment (ed1d1a8d) [reply building/build-debs/homeworld-admin-tools/inner-build.sh]: I am writing it to src/resources/GIT_VERSION, which should already be cleaned up by the existing code. Though this is not done with a trap. Should I change it to use a trap?

File: building/build-debs/homeworld-admin-tools/resources/__init__.py (hunk 10 above)
Comment (ed1d1a8d) [reply building/build-debs/homeworld-admin-tools/resources/__init__.py]: Moved to separate commit.

File: building/build-debs/homeworld-admin-tools/inner-build.sh (hunk 12 above)
Comment (celskeggs) [reply building/build-debs/homeworld-admin-tools/inner-build.sh]: Woops, you're right. It's good as it is.

**Comment by celskeggs** [comment]
Actually, this needs to be rebased. Sorry 'bout that.

**Comment by ed1d1a8d** [comment]
Rebased.
